testsuite/integration-arquillian/servers/auth-server/services/testsuite-providers/src/main/java/org/keycloak/testsuite/federation/BackwardsCompatibilityUserStorage.java

```diff
@@ -157,7 +159,7 @@ public boolean updateCredential(RealmModel realm, UserModel user, CredentialInpu
 
             users.get(translateUserName(user.getUsername())).hashedPassword = newPassword;
 
-            UserCache userCache = session.userCache();
+            UserCache userCache = UserStorageUtil.userCache(session);
```

Comment:
Good catch, In this case it is sadly not possible, and that is for the reason that user storage SPI will only be supported in legacy store and eventually removed. We aimed to keep the changes minimal, but this one is necessary.

I agree that this needs to be described in upgrading (migration) guide.

This however revealed that `DefaultKeycloakSession.userCache()` still returns the `UserCache` which should be moved to legacy module.

@ahus1 Could you please move it accordingly?

Reply:
Hi @hmlnarik - I had a look and `UserCache` is part of the interface of `KeycloakSession`, so I suppose it can be only deprecated in this release and not moved or removed (the same way we did this for `userCredentialManager()`).

And it also has the semantic that it returns null when no cache is available, and non-null when there is a cache. As there is no cache in map store mode, it will return null and it fine.

Reply:
`UserCache` is a `UserProvider`. Could you please deprecate it and make it return `null` by default? The existing code should be handled similarly to other `KeycloakSession` methods which were transformed in this PR.

Reply:
After talking to @hmlnarik I now know that I'm allowed to make this breaking change. 

While the method in `KeycloakSession` wasn't used any more, there were other classes and interfaces depending on `UserCache`, which have been refactored and moved. See 0f0d9fd11738a1beb37d6e81e5c5d08d9b5b91c4 for the commit.

Reply:
Migration guide updated: https://github.com/keycloak/keycloak-documentation/commit/4b46e0b0f27a93885b649d6a33f5a3a7506258ea

Next up: moving CacheRealmProvider as well.

Reply:
CacheRealmProvider also migrated: bfee3fad1024a7b3464a9522ab9961199673a633

No updates to the migration guide, as this was a private SPI.

Reply:
@mposolda The migration of custom user storage implementation is now documented in keycloak/keycloak-documentation#1541. Could you please check if that is sufficient from your perspective?

Reply:
@hmlnarik @ahus1 Approved the migration guide changes. Thanks!